core/drivers/crypto/hisilicon/hpre_montgomery.c

```diff
@@ -0,0 +1,508 @@
+{
+	uint32_t size = HPRE_MONTGOMERY_TOTAL_BUF_SIZE(msg->key_bytes);
+
+	msg->key = calloc(1, size);
+	if (!msg->key) {
+		EMSG("Fail to alloc montgomery key buf");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->key_dma = virt_to_phys(msg->key);
+	msg->in = msg->key + HPRE_X_KEY_SIZE(msg->key_bytes);
+	msg->in_dma = msg->key_dma + HPRE_X_KEY_SIZE(msg->key_bytes);
+	msg->out = msg->in + msg->key_bytes;
+	msg->out_dma = msg->in_dma + msg->key_bytes;
+
+	return 0;
+}
+
+static void hpre_montgomery_params_free(struct hpre_montgomery_msg *msg)
+{
+	if (msg->key) {
+		memzero_explicit(msg->key, HPRE_X_KEY_SIZE(msg->key_bytes));
+		free(msg->key);
+		msg->key = NULL;
+	}
+}
+
+static int32_t
+hpre_montgomery_params_pretreatment(struct hpre_montgomery_msg *msg)
+{
+	uint8_t *p = msg->key;
+	uint8_t *a = p + msg->key_bytes;
+	uint8_t *k = a + msg->key_bytes;
+	uint8_t *u = msg->in;
+	uint8_t *dst = k;
+	uint32_t bsize = msg->key_bytes;
+	uint32_t dsize = msg->curve_bytes;
+	uint32_t psize = msg->curve_bytes;
+	uint32_t shift = 0;
+	int32_t ret = 0;
+
+	/*
+	 * This is a pretreatment of X25519/X448, as described in RFC 7748:
+	 * For X25519, in order to decode 32 random bytes as an integer
+	 * scaler, set the three LSB of the first byte and MSB of the last
+	 * to zero, set the second MSB of the last byte to 1.
+	 * For X448, set the two LSB of the first byte to 0, and MSB of the
+	 * last byte to 1. Decode in little-endian mode.
+	 * HPRE hardware module uses big-endian mode, so the bytes to be
+	 * set are reversed compared to RFC 7748:
+	 * For example, dat[0] of X25519 in RFC 7748 is reversed to dat[31]
+	 * in HPRE specification, so does X448.
+	 * For X25519, when receiving u-array, set MSB of last byte to zero.
+	 * HPRE hardware module uses big-endian mode, so we set MSB of first
+	 * byte to zero instead.
+	 */
+	if (msg->key_bytes == BITS_TO_BYTES(X25519_KEY_BITS)) {
+		dst[31] &= 248;
+		dst[0] &= 127;
+		dst[0] |= 64;
+		u[0] &= 0x7F;
+	} else {
+		dst[55 + bsize - dsize] &= 252;
+		dst[0 + bsize - dsize] |= 128;
+	}
+
+	shift = bsize - p_size;
+	ret = memcmp(u + shift, p + shift, p_size);
+	if (ret >= 0) {
+		EMSG("u > p");
+		return HISI_QM_DRVCRYPT_EINVAL;
+	}
+
+	return 0;
+}
+
+static int32_t hpre_montgomery_params_fill(struct hpre_mgm_curve *curve,
+					   struct hpre_montgomery_msg *msg,
+					   uint8_t *privkey, uint8_t *pubkey)
+{
+	uint8_t *p = msg->key;
+	uint8_t *a = p + msg->key_bytes;
+	uint8_t *k = a + msg->key_bytes;
+	uint8_t *x = msg->in;
+	int32_t ret = 0;
+
+	memcpy(p, curve->p, msg->curve_bytes);
+	memcpy(a, curve->a, msg->curve_bytes);
+	memcpy(k, privkey, msg->curve_bytes);
+	msg->x_bytes = msg->curve_bytes;
+	if (!pubkey)
+		memcpy(x, curve->x, msg->x_bytes);
+	else
+		memcpy(x, pubkey, msg->x_bytes);
+
+	ret = hpre_bin_from_crypto_bin(p, p, msg->key_bytes, msg->curve_bytes);
+	if (ret) {
+		EMSG("Fail to transfer montgomery p from crypto_bin to
```

Comment:
Please keep this print in one line, same goes the for other prints. It's OK if this line with the format string becomes longer than 80.